src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -238,7 +238,18 @@ void RadioSpinel<InterfaceType, ProcessContextType>::Init(bool aResetRadio,
         SuccessOrDie(mSpinelInterface.ResetConnection());
     }
 
-    SuccessOrExit(error = WaitResponse());
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
+    do
+    {
+        RecoverFromRcpFailure();
+#endif
+
+        error = WaitResponse();
+
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
+    } while (mRcpFailed);
+#endif
```

Comment:
Needs `SuccessOrExit` in the case that recovery is disabled.

Better to log the error in the case that recovery is enabled

I'd suggest to use a single `#if` `#else` block.

Reply:
A second thought is that this PR may not be necessary. 

When `RCP RESTORATION` is enabled, the `WaitResponse()` won't actually return errors, so `Init` will continue the code below and still trigger RCP recovering if it failed in `WaitResponse`. 
Thoughts? @jinran-google

Reply:
`WaitResponse()` won't return errors, but `mIsReady` won't be set either. So it will still fail at the next line `VerifyOrExit(mIsReady, error = OT_ERROR_FAILED);`.

Reply:
That's a good point.

Reply:
refactored